adds saem convergence plot

djnavarro opened the pull request:
Continues work started by @certara-irebai on SAEM convergence plots in #382, fixing #152. SAEM convergence plot to display sequence of estimates for population parameters computed after each iteration of the SAEM algorithm from Monolix output

(work in progress)

djnavarro commented:
@mattfidler @A2P2 would it be possible to approve workflows on this while still in draft stage? tests will definitely fail but it would still be useful to see the CI results

mattfidler commented:
Hi @djnavarro  hopefully this works for you 😄

djnavarro commented:
thanks!

djnavarro commented:
@mattfidler @A2P2: This is not ready to merge, but close enough that review would be useful to check that this is on the right track. There are more changes than I'd like for a single PR: I wanted to ensure all tests are passing locally without warnings or errors, so this has sprawled more than would be ideal

A2P2 commented:
@djnavarro Thanks! I'l try to review by Wednesday 26.03. Tests failed due to variables being passed to ggplot2 and dplyr. Shall `aes_string` be used instead of `aes`? For dplyr I think you could add `.data$` in front of the variable name.

djnavarro commented:
@A2P2 I pushed a few quick fixes that should clean up some (but not all) of the warnings and notes from R CMD check. The approach I took in regards to unquoted variable names passed to ggplot2 and dplyr was to declare them with `utils::globalVariables()`, but we can revisit later if a different approach makes more sense :)

A2P2 commented:
@djnavarro  Really good overall. The implementation into ggPMX looks correct.  Plot works for me and additional arguments work as well `ctr%>%pmx_plot_saem_convergence(filter = (Iteration >100))` (there is probably no good reason to filter like that, but good it works, confirming correct implementation)

Small things:
1. It doesn't apply properly plot config in some cases. `ggPMX::get_plot(ctr, 'saem_convergence')` returns a plot with smoothed red lines. `ggPMX::get_plot_config(ctr, 'saem_convergence')` returns config with `is.smooth = TRUE`

2. Shall we add additional argument ncol? 
From monolix R api:

>  a list of optional settings:
> grid (logical) add (TRUE) / remove (FALSE) plot grid (default TRUE).
> ncol (integer) number of columns when facet = TRUE (default 4).
> fontsize (integer) Plot text font size.

3. To check my understanding, there is no explicit test for the saem plot.  It tests if it can be plotted in general in `test-pmx-all-plots.R`, right? Does it make sense to extend tests or good as is?

djnavarro commented:
@A2P2 This should be ready for re-review. Update addresses all issues above:

1. Fixed. Parameters were being passed incorrectly to `pmx_gpar()`
2. Agreed. The update now does this by passing a `facet` argument (similar to what is done with the individual subject plots)
3. You are correct, I hadn't implemented any tests specifically for the new convergence plot. I've now added a minimal test suite that checks plots can be produced, that arguments like `filter` work, and that the plot can be styled correctly
4. Done. 
5. In addition to the above, I've also restructured the SAEM plot so that the colour scheme is no longer hard coded and can be specified through the YAML file. It also allows the phase 2 reference line to be removed or styled separately 

Tests are passing locally, and R CMD check seems good too.

mattfidler commented on NAMESPACE:
```diff
@@ -1,13 +1,15 @@
 # Generated by roxygen2: do not edit by hand
 
 S3method("[",pmx_gpar)
+S3method(base::print,classIntervals)
```
Should this be exported?  I think it shouldn't be exported

mattfidler replied:
Actually it likely should be exported, but the "classIntervals" `s3` should be changed since `classInt` is  a popular package and is likely to cause issues with many other packages.

mattfidler commented on inst/init/standing.ppmx:
```diff
@@ -510,3 +520,30 @@ PMX_VPC:
+
+## SAEM plots
+    
+SAEM_CONVERGENCE:
```
While this can be use for `saem_convergence` some people also adjust and look at `focei`; perhaps the name can be generalized;  In `saem` this is a convergence plot, but in other methods it is a parameter history plot.

mattfidler replied:
Sounds good to me.

mattfidler reviewed (changes requested):
This seems OK with me, with a minor comment

- `classIntervals` is a class in a popular package `classInt`, so I think it should be renamed.

What do you think?

djnavarro commented:
Hi @A2P2, hi @mattfidler - sorry it has taken me so long to return to this (as mentioned in various other places, I've been out of action with covid for a while). I've made some changes to address one of Matt's suggestions (the plot type renaming), and one other minor tweak to stop ggplot2 throwing deprecation warnings when `is.ggplot()` is called. As mentioned above I agree that we should fix the issue associated with `classInt` namespace conflicts asap, but suggest that be the next PR after merging this one: I'd like to see what happens if we take a dependency on `classInt`, but wary of doing it at the tail end of this long PR

djnavarro commented:
> @djnavarro do we need to regenerate snapshots?

@A2P2 we might? I confess this is the first time I've dealt with vdiffr snapshots in tests so I'm not certain when they need to be re-generated. I'll take a look - it's high time I wrapped my head around snapshot tests anyway :)  

> One note @djnavarro and @A2P2 about adding `classInt`, I think we are close to the 20 imports that CRAN limits packages to. Some creative Suggests may be required

@mattfidler Agreed. Happy to take a look at that

mattfidler commented:
> @A2P2 we might? I confess this is the first time I've dealt with vdiffr snapshots in tests so I'm not certain when they need to be re-generated. I'll take a look - it's high time I wrapped my head around snapshot tests anyway :)

If there are any differences, you manually check to see if they impact the visualization.

Since `ggplot` has been completely rewritten, you may want to test against that version if you decide to do this now.  My guess is a major release version is likely to change the order of some components in the SVG (ie drawing order), so you would have to do this again after that point.

https://github.com/nlmixr2/nlmixr2plot/issues/39

Has instructions on how to install the development version of ggplot2

djnavarro commented:
temporarily disabled vdiffr snapshot tests on github and cran, since the ggplot2 versions we're using locally vs on CI are substantively different. none of the diffs seemed very troubling so perhaps best to wait until ggplot2 4.0.0 is released before trying to resolve

@A2P2 - I've updated the `is_ggplot()` function using `getNamespaceExports()` to detect if the underscore version is present

I'd be happy to merge this version and then start fresh in a new PR with some of the remaining issues

A2P2 commented:
Impressive 13k lines of code
